### ISSUE TITLE
Some LZOP archives don't work

### ISSUE THREAD

**Issue by fdr** [wal-g/wal-g#35]
I'm still taking this apart but it can be related to #22.

I have this backup that has hundreds of archives, but one of those archives causes a systematic crash, whereas `lzop` seems fine with it. There's not much to do but for me to go through it with a fine-tooth comb, but, FYI.

cc @x4m

**Comment by x4m** [comment]
Is this just one file? Can you share it with me? I'd like take a look.

**Comment by fdr** [comment]
On Mon, Oct 30, 2017 at 10:19 PM Andrey Borodin <notifications@github.com>
wrote:

> Is this just one file? Can you share it with me? I'd like take a look.
>

Nope. Confidential data.

**Comment by fdr** [comment]
Okay. It's a bug, and not the first one, in the rasky pure-go lzo library. The C binding version works perfectly. At this point I have to decide whether I want to involve a C dependency (probably making it optional via build tag) or fix the rasky go-lzo library.

**Comment by fdr** [comment]
I think I'm going to go the route of binding the C lzo library via build tag. LZO support is basically transitional anyway, and sinking time into fixing the pure-go implementation may not be worth it.

**Comment by PierreDucroquet** [comment]
I'm facing a similar issue. I tried replacing wal-e by wal-g for a restore job and got issues with some wal files that could not be uncompressed by wal-g while lzop accept them happily.
I can not share the file for privacy/legal reasons, but I would be happy to help testing new code if needed.

**Comment by x4m** [comment]
@PierreDucroquet , do you use GPG encryption? Can you share error message?
If you have a message "DecompressLzo: did not fill skip", probably, it is another issue.

Just today I've encountered an error, which manifests oftener with GPG encryption: here's the call https://github.com/wal-g/wal-g/blob/master/decompress.go#L46 which must use ioutil.ReadFull(). Hope to fix this week.

**Comment by PierreDucroquet** [comment]
Yep, we have GPG encryption and it had another issue I worked around.

Basically, my patched function looks like : 
```go
func DecompressLzo(d io.Writer, s io.Reader) error {
        skip := 33
        sk := make([]byte, skip)

        n, err := s.Read(sk)
        full_skip := n
        if n < skip {
                for {
                        _sk := make([]byte, skip - full_skip)

                        n, err = s.Read(_sk)
                        full_skip += n
                        if full_skip == skip {
                                break
                        }
                }
        }
        if full_skip != len(sk) {
                return errors.New(fmt.Sprintf("DecompressLzo: did not fill skip, got only %d bytes -- %s", full_skip, sk))
        }
        if err != nil {
                return errors.Wrap(err, "DecompressLzo: read failed")
        }
```

But since I could not fully test that it works (since I had that second LZO issue), I chose not to submit it (also, I never wrote Go before, and to be honest I don't like that language very much and feel not comfortable at all with it)

**Comment by x4m** [comment]
@PierreDucroquet may I ask you to test this patch?
https://yadi.sk/d/ykc-9CYf3RXpdv

**Comment by x4m** [comment]
@PierreDucroquet you can test it in current pre-release.

**Comment by PierreDucroquet** [comment]
I confirm, it fixes the “DecompressLzo: did not fill skip“ error message. Only the bug in the pure-go lzo library remains.

**Comment by x4m** [comment]
Thanks!
Unfortunately, I'm not one of those happy(?) owners of broken lzo file.. can't fix this

**Comment by PierreDucroquet** [comment]
@x4m I was under the impression you were going to switch from the pure go lzo library to a wrapper around the C lzo library. If it's not what you intend to do, I could then try to spend some time on this issue (I already have a small unlzo binary written using the go library to confirm the issue)

**Comment by x4m** [comment]
@PierreDucroquet well, this topic is in @fdr's hands.
I just think that it's not pointless to fix that library regardless of it's usage in WAL-G, probably it is used elsewhere and their users suffer from this bug too.

**Comment by jonjomckay** [comment]
I'm also suffering from the problem of one of our archives causing the `backup-fetch` to fail (right at the end, annoyingly!). Has there been any success in swapping the lzo library out?

**Comment by x4m** [comment]
@jonjomckay  I'm not working on this, sorry. I have no broken lzo file to work on this and, probably, will never have - I've migrated to WAL-G which uses lz4. Maybe @fdr will fix this eventually...

**Comment by erkkah** [comment]
I can just confirm that this is happening for us as well.
We are using WAL-E for replication, which makes it problematic to make the switch to lz4 directly.

**Comment by x4m** [comment]
This issue was fixed in #100 and fix is released.
Feel free to report new issues if you observe some related misbehavior.
